### ISSUE TITLE
Should drush continue to support the Windows platform?

### ISSUE THREAD

**Issue by greg-1-anderson** [drush-ops/drush#1612]
Windows is not tested under CI, and it seems that Windows-related Drush questions do not get good support, either in this queue or on Stack Exchange.  This situation has gone on for years.

Roman Paska claims that [Drush on Windows is mostly functional](https://mobile.twitter.com/T2LPR/status/642041666208178177), although we still need a little more information about which shells work, what versions of Drush folks are using, etc.

I think that if we could set up CI for Windows Drush, and had a volunteer to do that, and if we had a number of people who could post here and say, "My name is ____, I use Drush on Windows regularly, and I have subscribed to the #drush tag on Drupal Stack Exchange, and will answer Drush on Windows questions there", well, then we'd have a supported platform, and there would be no question about that remaining.

I am somewhat dissatisfied with the current "unsupported but happens to work" state of affairs, though, and would prefer to firmly be in one camp (supported) or the other (support removed).

**Comment by jibran** [comment]
How about this:

1. Find a good maintainer for windows.
2. Move the windows hacks out of drush.
3. Create a new project drush-win for windows only.

**Comment by jibran** [comment]
Found https://github.com/travis-ci/travis-ci/issues/216 & https://github.com/travis-ci/travis-ci/issues/2104 after quick search.

**Comment by weitzman** [comment]
I do agree that setting up CI for Windows is the crucial missing piece. Travis is never going to support Windows AFAICT. I've been subscribed to those issues for a long time.

Looks like http://www.appveyor.com/docs/build-configuration is free for open source projects and supports Windows build machines. We could run it in parallel with Travis.

Would be great if someone forked Drush and showed it running on appveyor. All you have to do is sign up on appveyor and add an appveyor.yml to your fork.

**Comment by weitzman** [comment]
I got pretty far along with automated testing on Appveyor. Drush is now installed via Composer. Now we need to get Unish (our test suite) to complete a test run. See https://ci.appveyor.com/project/weitzman/drush/build/1.0.53/job/s70c2tla8fmoppu8 for the current trouble clearing the sandbox and finding drush.bat.

Help appreciated.

**Comment by greg-1-anderson** [comment]
Great progress!  From the output, it looks as if the permission denied is caused by Unish trying to remove the sandbox on the first run, before it exists.  The code is wrapped in 'file_exists', though, so I don't know what to make of that.

My guess about the drush.bat is it seems that there are too many \'s in the PATH -- but I'm not sure about that.

**Comment by weitzman** [comment]
A bit more hacking and the tests are now running. I added a little debugging to solve the Sandbox delete problem but it's not licked yet - https://ci.appveyor.com/project/weitzman/drush/build/1.0.68/job/0ey7dttaiqi75u4w

**Comment by weitzman** [comment]
Tests are proceeding now but Drupal won't install despite my seemimgly providing valid creds. The same creds are working earlier in the process. Once we lick this problem, we need to fix all the quoting, slashes, etc. so that tests pass. I'll be asking for help on this. You can see where we are at by reviewing https://ci.appveyor.com/project/weitzman/drush/build/1.0.92

Sorry for the barrage of commits on appveyor branch. In hindsight, I should have forked before starting this.

**Comment by greg-1-anderson** [comment]
Fantastic.  Sorry I haven't had time to help get the test harness set up; I'll help with quoting and escaping issues once we get to that point.

It would be nice if there was someone from the Windows community who could help out here.  Not much demand for Drush on Windows, maybe?

**Comment by mikeker** [comment]
I joking call myself the last core contributor on Windows. I realize that's an exaggeration -- there are likely others and I've only contributed to a dozen or so core issues. But I do dev work on Windows and use Drush every single day. I could not live without it! I would hate to see Drush give up on Windows support especially since Drupal and PHP both support Windows as first-class citizens.

As requested, "My name is Mike Keran, I use Drush on Windows regularly, and I have subscribed to the #drush tag on Drupal Stack Exchange, and will (as best I can) answer Drush on Windows questions there." Though I hardly use SE so it may take me a while to get into the swing of things.

Perhaps a better choice is to limit Windows support to the MinGW shell (which comes bundled with Git for Windows, so most dev should have it installed?) Dropping PowerShell support would remove the need for parallel .bat files. At that point, there's just rsync and a couple of oddball GnuWin32 commands and the / vs \ path delimiters to deal with. Or am I over-simplifying things?

Regardless, there is a clear need for CI on Windows as #1595 just broke Drush on Windows. PR coming soon...

**Comment by das-peter** [comment]
I totally agree with mikeker, I created patches for drush years ago since some stuff didn't work on Windows. But back then interest on that work wasn't really there.
Since then I used drush with MinGW on a daily base.

And I found this issue too while looking what was up with the latest drush on Windows because of: #1595 ;)

**Comment by weitzman** [comment]
@mikeker and @das-peter  - thanks so much for chiming in here. We would love your help with Drush on Windows. I would love help getting the Drush test suite to work on AppVeyor, as per this issue. I think your best bet is to fork Drush and hack away in its appveyor branch. You will want to setup a new project on [http://www.appveyor.com/](http://www.appveyor.com/) and point it at your fork.

Feel free to ask questions here or in #drush in IRC. When you can get Drush to install itself during the test suite, submit a new PR or announce it here.

**Comment by weitzman** [comment]
@mikeker, @das-peter  - Are you guys interested and available in getting Appveyor going? I'm happy to talk you through it by phone if that would avoid some startup headaches.

**Comment by mikeker** [comment]
@weitzman, I think it would be best if I do some basic research and try setting things up before I waste your time on the phone. I'll try to do that this week. I'm new to Drush dev and CI systems in general so there's a bit of a learning curve for me...

**Comment by das-peter** [comment]
@weitzman Thanks for the offer! I've the same perspective as @mikeker and currently I'm veeery busy working on a project while in parallel moving from CH to the US and then to NZ. Thus I've just little time I can spare for all the Drupal stuff :|

**Comment by cweagans** [comment]
Just to clarify here - is the problem Windows itself? Or the crappy "shell" it ships with? Would simply requiring a decent shell be a good solution here? I've heard great things about Babun on Windows, which is just zsh, cygwin, and some other nice things.

I know this has been discussed to death already, but would using something like the Console component or Climate to abstract away all of the weird crap that Windows does be a viable solution here? Composer seems to run well enough on Windows, so I was thinking that maybe the fact that it's a Console app is how they've avoided dealing with Windows headaches.

**Comment by mikeker** [comment]
> Just to clarify here - is the problem Windows itself? Or the crappy "shell" it ships with?

A little bit of both, in my opinion. Windows has some oddities (slashes, MAX_PATH, different file permissions) and cmd.exe is completely useless. I think we can work around the oddities as long as we have a relatively *nix-ish environment to work in.

Personally, I use the MinGW shell that ships with Git and a small pile of GnuWin32 utils to give me essentially all I need to minic *nix commands on Windows. I never use cmd.exe.

I think requiring a decent shell (MinGW, perhaps Cygwin -- never used it so I can't speak to it) is a reasonable requirement. I mentioned that earlier in this thread, but didn't hear a reply from the maintainers on that. It could turn off a CLI novice by requiring them to install a bunch of prerequisites...

**Comment by cweagans** [comment]
> It could turn off a CLI novice by requiring them to install a bunch of prerequisites...

They'll have to do that anyway at some point, I think. Having a reasonable shell is a requirement for most major open source CLI applications. I don't know of any exceptions to that (which is not to say that there aren't any - just that I don't know about them). It's very seldom that I use Windows for much of anything, but when I do, I use babun just because it's fast and easy to get up and running. That's not to say that Drush should require Babun, per se - just that it might be a good thing to recommend in the installation docs for Windows for people that doesn't know/want to know how to configure some other shell, or even how to choose something different. Better to just recommend something that is known to work in a lot of configurations.

**Comment by mikeker** [comment]
I say this having never tried Babun -- though I'm off to read up on it right now...  :)

>Better to just recommend something that is known to work in a lot of configurations.

In my opinion that would be MinGW, the shell that comes with Git (witth one minor tweak that we can document -- bsdtar vs. tar). If we can assume that CLI novices are biting the CLI bullet, I believe we can assume they'll be biting the Git bullet as well. The vast majority of Drupal dev work is centered on Git so we can safely assume that they have (or will soon have) Git installed if they're making the dive into the CLI.

**Comment by kenorb** [comment]
`drush` works on Windows fine (with Git shell installed), however the only problem which I've seen is that by default it's trying to run `drush.launcher`, which is doesn't work properly, so the workarounds are:

- [write a wrapper script](http://superuser.com/q/998180/87805)
- if installed via Composer, use `drush.launcher.bat` instead (or `alias drush=drush.php.bat`)
- [install via `drush.phar`](https://github.com/drush-ops/drush/issues/1696#issuecomment-155575706)

So there are few minor tweaks to work on.

Writing [appveyor](http://www.appveyor.com/) tests is actually good idea (+ additional dependencies installed via [`choco`](Chocolatey)).

**Comment by hansfn** [comment]
"My name is hansfn, I use Drush on Windows regularly, and I have subscribed to the #drush tag on Drupal Stack Exchange, and will answer Drush on Windows questions there."

Regarding the shell: I have used Mingw in various packaging for years, but I think we should support Powershell as the default/preferred shell on Windows. ("GitHub Desktop" or "GitHub for Windows" does exactly that for Git with Mingw (Git Bash) as an optional shell.) Let's just forget about Cmd.

So what can I do to help you guys?

**Comment by weitzman** [comment]
Thanks @hansfn. From my perspective, the best bext step is for someone to get the tests passing (or at least completing) on Appveyor. Read up on this issue for details. I'm happy to help when you guys get stuck.

**Comment by Benosika** [comment]
All the questions arising here about Drush-master windows intergration shows how big is the demand for using it on Windows. The Drupal community is raising to new heights. The "Drupal support" group on Facebook grew from 4000 friends to 6500 in 3 months. The growing number of high-quality documentation (High-quality == Not only for Devs\Programmers) is a main reason for that, and if we enlarge the community even more by accessing it to more people, than I think Drush For Windows is an integral part of it, and if I had enough knowledge in Programming I would gladly do so; Right now I just now some fundamental JS and Can barely code with it.

I am an SEOer and A Drupalist who loves using Drush, mostly because I love to save time and work efficiently, but for other reasons as well. I am also a MS Windows user for at about 20 years now, since there was Windows 3.1 ... I love this OS very much and love it more each time a new version comes out... Windows 10 for example is a very fast OS, The fastest of them all, by all means.

Drush also saved me several times when problematic modules\themes\updates\whatever crashed a site. I have no intention of not using it.

I write this text from an Ubuntu-based environment that I am about to remove because of many problems I encounter not because of lazyness (believe it or not, I'm not trying to convince anyone) but because of performance\drivers\programs-Interface\language\keyboard-syncronization\compatibility problems that are not that fast to solve, and I don't have the resources to solve that easily at least in the coming year.

So my idea is that removing Drush's Windows support is a not only a bad idea but also a dangerous one, and if there are alternatives like Git For Windows, they should be mentioned in the same thread\comment, wherever on the net.

I think that if there isn't one already, there should be a software package for Drush --- Unix based software a user needs to run Drush stable and fast on Windows - With a "single" mouse click.

**Comment by jibran** [comment]
Here is a start https://github.com/jibran/drush/pull/1.

**Comment by christopher-hopper** [comment]
@greg-1-anderson, 
On Drush for Windows there is some great news from Microsoft:

> Windows provides developers with a familiar Bash environment. 
> 
> Windows is running Ubuntu user-mode binaries provided by Canonical. This means the command line utilities are the same as those that run within a native Ubuntu environment.

You can now install a bash command-line on Windows, that runs in an Ubuntu-like environment, which is fully supported by both Microsoft and Canonical. 

- [About Windows Subsystem for Linux](https://msdn.microsoft.com/commandline/wsl/about)
- [Windows Subsystem for Linux Installation Guide](https://msdn.microsoft.com/en-gb/commandline/wsl/install_guide)

I've been using Drush on Windows for many years by installing [Cygwin](http://cygwin.org/) with some packages (like PHP) provided by [Cygwin Ports](http://cygwinports.org/). This new development at Microsoft changes the game however. 

As a first step, I'd like to update the Install Alternatives documentation to reflect this new development from Microsoft.

**Comment by christopher-hopper** [comment]
Okay, I've been reading up on http://www.appveyor.com/ and the progress so far in running tests. 

I'll have a crack at getting something started using the new Windows Subsystem for Linux and post back here if I get something working.

**Comment by weitzman** [comment]
@christopher-hopper - that would be a huge service for many people. Thanks!

**Comment by hansfn** [comment]
Just a minor remark: It seems "Windows Subsystem for Linux" only is available for recent versions of Windows 10. Some of us are still stuck on Windows 7 :-(

**Comment by christopher-hopper** [comment]
For Windows 7 I'd stick with Cygwin and Cygwin Ports as the best experience I've ever achieved with Drush. If I can get the WSL bash tests running first, then maybe I'll look at how to automate tests and document an install using Cygwin. 

As for the Windows Cmd.exe, that's just not worth the effort anymore.

**Comment by Benosika** [comment]
So happy I could finally use drush-master smoothly on Win10 ...

**Comment by Benosika** [comment]
Someone who managed to run Drush 8 / Master on Win10 with WSL might want to share the steps taken for that here:

http://drupal.stackexchange.com/questions/201229/is-it-possible-to-run-drush-8-smoothly-on-win10-wsl-windows-subystem-for-linux

**Comment by Benosika** [comment]
I installed WAMP, WSL and Drush on WSL ... I than navigated to a site's folder in WAMP with WSL bash like this:

```
cd /
cd mnt/c/wamp64/www/yoursitesname/
```
While inside the sites' folder I did "drush up" and than got:

```
Benia@DESKTOP-D5K6365:/mnt/c/wamp64/www/benia$ drush up
PHP Fatal error:  Undefined class constant 'MYSQL_ATTR_SSL_CA' in phar:///usr/bin/drush/lib/Drush/Sql/Sqlmysql.php on line 56
Drush command terminated abnormally due to an unrecoverable error.                                                                                 [error]
Error: Undefined class constant 'MYSQL_ATTR_SSL_CA' in phar:///usr/bin/drush/lib/Drush/Sql/Sqlmysql.php, line 56
Benia@DESKTOP-D5K6365:/mnt/c/wamp64/www/benia$
```

**Comment by kenorb** [comment]
@Benosika You should report a separate bug for this.

**Comment by weitzman** [comment]
Still looking for a hero to pick up this effort.

**Comment by weitzman** [comment]
Good news!

I finished a bunch of Windows compatibility work on the Finder which will make life much better. See #2312.

[We are now successfully running all tests at Appveyor](https://ci.appveyor.com/project/weitzman/drush). There are still plenty of fails but many need fixes in tests, not in the Drush code. Please open PRs against Drush's [appveyor branch](https://github.com/drush-ops/drush/tree/appveyor) for fixing these. Hopefully Appveyor wlll test those PRs.

**Comment by greg-1-anderson** [comment]
Just a note -- The various consolidation projects (robo, annotated-command, output-formatters and log) are now completely tested in Appveryor, so there shouldn't be any further failures due to deficiencies in these upstreams.
